File: gareth-api/src/main/java/org/craftsmenlabs/gareth/rest/RestErrorHandler.java
package org.craftsmenlabs.gareth.rest;

import org.springframework.http.client.ClientHttpResponse;
import org.springframework.web.client.DefaultResponseErrorHandler;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.util.Collection;
import java.util.LinkedList;
import java.util.StringJoiner;

public class RestErrorHandler extends DefaultResponseErrorHandler {

    private boolean hasError = false;
    private String responseCodeTxt = null;
    private int responseCode;
    private String responseBody;

    public void handleError(ClientHttpResponse response) throws IOException {
        hasError = true;
        responseCodeTxt = response.getStatusText();
        responseCode = response.getRawStatusCode();
        Collection<String> strings = readLines(response.getBody());
        StringJoiner stringJoiner = new StringJoiner("");
        strings.forEach(s -> stringJoiner.add(s));
        responseBody = stringJoiner.toString();
    }

    public String getResponseCodeTxt() {
        return responseCodeTxt;
    }

    public int getResponseCode() {
        return responseCode;
    }

    public boolean hasError() {
        return hasError;
    }

    public String getResponseBody() {
        return responseBody;
    }

    private Collection<String> readLines(InputStream is) {
        LinkedList out = new LinkedList();
        try {
            BufferedReader r = new BufferedReader(new InputStreamReader(is, Charset.forName("UTF-8")));
            String line;
            while ((line = r.readLine()) != null) {
                out.add(line);
            }
            return out;
        } catch (IOException ioe) {
            throw new RuntimeException("Problems reading from: " + is, ioe);
        }
    }

    public void reset() {
        hasError = false;
        responseCodeTxt = null;
        responseCode = 0;
        responseBody = null;
    }
}

File: gareth-api/src/main/kotlin/org/craftsmenlabs/gareth/rest/RestErrorHandler.kt
package org.craftsmenlabs.gareth.rest

import org.springframework.http.client.ClientHttpResponse
import org.springframework.web.client.DefaultResponseErrorHandler
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.charset.Charset
import java.util.*


class RestErrorHandler : DefaultResponseErrorHandler() {

    var hasError = false
        private set
    var responseCodeTxt: String? = null
        private set
    var responseCode: Int = 0
        private set
    var responseBody: String? = null
        private set

    override fun handleError(response: ClientHttpResponse) {
        hasError = true
        responseCodeTxt = response.statusText
        responseCode = response.rawStatusCode
        val strings = readLines(response.body)
        val stringJoiner = StringJoiner("")
        strings.forEach { s -> stringJoiner.add(s) }
        responseBody = stringJoiner.toString()
    }

    private fun readLines(inputStream: InputStream): List<String> {
        try {
            val r = BufferedReader(InputStreamReader(inputStream, Charset.forName("UTF-8")))
            return if (r != null) r.readLines() else listOf()
        } catch (ioe: IOException) {
            throw RuntimeException("Problems reading from: " + inputStream, ioe)
        }
    }

    fun reset() {
        hasError = false
        responseCodeTxt = null
        responseCode = 0
        responseBody = null
    }
}